ext/BijectorsEnzymeCoreExt.jl

```diff
@@ -0,0 +1,184 @@
+module BijectorsEnzymeCoreExt
+
+if isdefined(Base, :get_extension)
+    using EnzymeCore:
+        Active,
+        Const,
+        Duplicated,
+        DuplicatedNoNeed,
+        BatchDuplicated,
+        BatchDuplicatedNoNeed,
+        EnzymeRules
+    using Bijectors: find_alpha
+else
+    using ..EnzymeCore:
+        Active,
+        Const,
+        Duplicated,
+        DuplicatedNoNeed,
+        BatchDuplicated,
+        BatchDuplicatedNoNeed,
+        EnzymeRules
+    using ..Bijectors: find_alpha
+end
+
+# Compute a tuple of partial derivatives wrt non-`Const` arguments
+# and `nothing`s for `Const` arguments
+function ∂find_alpha(
+    Ω::Real,
+    wt_y::Union{Const,Active,Duplicated,BatchDuplicated},
+    wt_u_hat::Union{Const,Active,Duplicated,BatchDuplicated},
+    b::Union{Const,Active,Duplicated,BatchDuplicated},
+)
+    # We reuse the following term in the computation of the derivatives
+    Ωpb = Ω + b.val
+    c = wt_u_hat.val * sech(Ωpb)^2
+    cp1 = c + 1
+
+    ∂Ω_∂wt_y = wt_y isa Const ? nothing : oneunit(wt_y.val) / cp1
+    ∂Ω_∂wt_u_hat = wt_u_hat isa Const ? nothing : -tanh(Ωpb) / cp1
+    ∂Ω_∂b = b isa Const ? nothing : -c / cp1
+
+    return (∂Ω_∂wt_y, ∂Ω_∂wt_u_hat, ∂Ω_∂b)
+end
+
+# `muladd` for partial derivatives that can deal with `nothing` derivatives
+_muladd_partial(::Nothing, ::Const, x::Union{Real,Tuple{Vararg{Real}},Nothing}) = x
+_muladd_partial(x::Real, y::Duplicated, z::Real) = muladd(x, y.dval, z)
+_muladd_partial(x::Real, y::Duplicated, ::Nothing) = x * y.dval
+function _muladd_partial(x::Real, y::BatchDuplicated{<:Real,N}, z::NTuple{N,Real}) where {N}
+    let x = x
+        map((a, b) -> muladd(x, a, b), y.dval, z)
+    end
+end
+_muladd_partial(x::Real, y::BatchDuplicated, ::Nothing) = map(Base.Fix1(*, x), y.dval)
+
+function EnzymeRules.forward(
+    ::Const{typeof(find_alpha)},
```

Comment:
this is using the rule system for Enzyme <=0.12, which was changed in Enzyme 0.13. [or equivalently EnzymeCore 0.7 and 0.8 iirc]. We should probably use the latter.

The breaking change of relevance is that forward mode rules now have a config as the first arg [similar to ReverseMode], with utility functions that specify whether the original return and/or derivative returns are requested.

For example: https://github.com/EnzymeAD/Enzyme.jl/blob/31df08b376634e2926e1d147c67284b20c493c22/src/internal_rules.jl#L858

Reply:
> this is using the rule system for Enzyme <=0.12, which was changed in Enzyme 0.13.

Yes, I know, this is intentional since the master branch uses Enzyme 0.12 and I wanted to keep the update to Enzyme 0.13 and the switch from the imported to the explicit rules separate.

Reply:
> The breaking change of relevance is that forward mode rules now have a config as the first arg [similar to ReverseMode], with utility functions that specify whether the original return and/or derivative returns are requested.

I saw this, and one thing that confused me was: What's the purpose of the return type/activity (Const/Duplicated/DuplicatedNoNeed) if it's solely based on `needs_primal(config)` and `needs_shadow(config)` what's returned? Why does the argument exist at all if it is ignored?

Reply:
For example, it was previously impossible to represent a function with a constant return, whose primal result was not required by enzyme

Reply:
Simultaneously marking a return DuplicatedNoNeed is no longer how to specify that the primal is not needed (eg outside of rules you would do autodiff(ForwardWithPrimal, Duplicated, …)

Reply:
Hmm, I think I still don't fully understand. Based on the `sort!` example it seems the only use case would be to dispatch on batched/non-batched mode? But that could be done solely by dispatching on argument `xs`, it seems? Both versions (batched/non-batched) completely ignore whether the return type is Const/Duplicated/...

Reply:
So that’s one use of it.

But basically they represent different (though related) pieces of information.

const:
  Result is not differentiated wrt.
  It can either be required to be returned (needs primal) or not
Duplicated:
  Result is differentiated wrt
  The primal can be required to be returned or not (previously this was represented in dupnoneed vs duplicated, now duplicated is always passed in?
  The shadow can be required to be returned or not (needs shadow). Prior to 0.13 it was always required to be returned, even if not needed.